### ISSUE TITLE
How to create local Identity User when using AAD?

### ISSUE THREAD

**Issue by Znow** [abpframework/abp#2815]
Hi,

As asked in a previous issue #2614 , I would like to request help on how to create a user in the DB, when signing in with AAD.

I have implemented the AAD logic in ProjectNameWebModule.cs "ConfigureAuthentication"-method, but how can in here, when "OnTokenValidated", create a user, if it does not exist in the DB?

So I can login a user when logging in with AAD? :-)

Thanks in advance,

Regards

**Comment by gterdem** [comment]
You are looking for [Signin with External Provider](http://docs.identityserver.io/en/latest/topics/signin_external_providers.html). Did you check this documentation?

**Comment by Znow** [comment]
@gterdem I have implemented signin with AAD... So what I need is to create a local user, when a user signs in with AAD, so I can buy the other login mechanism, enforced by ABP? :-)

**Comment by gterdem** [comment]
@Znow So you connected AAD as External OIDC provider (like google, facebook etc), seeing as external provider in your login page right?
When user selects Login with AAD (or whatever you named it), it will be redirected to AAD login page to sign in.
After user logs in to AAD Login page successfully;
If user is logging to your app first time, external registry screen will be displayed; asking email address to save (you can modify as you like). So your new user will be created on this step. 

Abp uses aspnet identity underneath so you can make any modifications you want in registry or login flow as in aspnet identity.

**Comment by Znow** [comment]
@gterdem Thanks alot - I can now see a "Azure AD login"-button on my Login page. Buuut, with correct setup in WebModule file, I am being thrown this error:

![image](https://user-images.githubusercontent.com/265719/74587735-9e9f2380-4ff6-11ea-89a5-235b62e70126.png)

I have seen this error several times before also with Azure AD integration

**Comment by hikalkan** [comment]
I suppose the problem is that AD sets user id claim which is obviously not a GUID and the ABP framework tries to read it as GUID.
You should somehow change this claim to another one (after return from the AD).

**Comment by Znow** [comment]
@hikalkan I have a total standard implementation, so can't see how I should or can change this? :-)

**Comment by hikalkan** [comment]
We can work on that to see details, but unfortunately now in this week (because we will release 2.1 in this week). If you have time, we can better help on this. I believe this is needed by many developers.

**Comment by Znow** [comment]
@hikalkan I can sure try to help, but please try and point me in some direction. I'm not that much into how Identity Server 4 works.

**Comment by gterdem** [comment]
@Znow Let's go step by step to troubleshoot and debug;
Here are some steps I tried to reproduce your issue; maybe it can help for you.

Create a page to override current login logic and override `CreateExternalUserAsync` method
```
[ExposeServices(typeof(LoginModel))]
public class MyLoginModel : IdentityServerSupportedLoginModel
{
        public MyLoginModel(IAuthenticationSchemeProvider schemeProvider, IOptions<AbpAccountOptions> accountOptions,
                            IIdentityServerInteractionService interaction, IClientStore clientStore,
                            IEventService identityServerEvents) 
            : base(schemeProvider, accountOptions, interaction, clientStore, identityServerEvents)
        {
        }
        protected override async Task<IdentityUser> CreateExternalUserAsync(ExternalLoginInfo info)
        {
            return base.CreateExternalUserAsync(info);
        }
}
```
Lets replace `return base.CreateExternalUserAsync(info);` with the source code
https://github.com/abpframework/abp/blob/037ef9abe024c03c1f89ab6c933710bcfe3f5c93/modules/account/src/Volo.Abp.Account.Web/Pages/Account/Login.cshtml.cs#L206-L217
And you can put a debugger to see which claim causes the issue.
```
protected override async Task<IdentityUser> CreateExternalUserAsync(ExternalLoginInfo info)
        {
            var emailAddress = info.Principal.FindFirstValue(AbpClaimTypes.Email);

            var user = new IdentityUser(GuidGenerator.Create(), emailAddress, emailAddress, CurrentTenant.Id);

            CheckIdentityErrors(await UserManager.CreateAsync(user));
            CheckIdentityErrors(await UserManager.SetEmailAsync(user, emailAddress));
            CheckIdentityErrors(await UserManager.AddLoginAsync(user, info));

            return user;
        }
```

**Comment by Znow** [comment]
@gterdem Thanks, I will look at it today or tomorrow and give you an update

**Comment by Znow** [comment]
@gterdem Could you provide an example of your source code? So I can double check the setup?

**Comment by Znow** [comment]
@gterdem Could you please provide an example of your source code? I have trouble with getting the method to be debugged.

![image](https://user-images.githubusercontent.com/265719/75097551-46df5a00-55ac-11ea-9e19-53fa6352e26e.png)

**Comment by gterdem** [comment]
@Znow Remember this will hit only once after returning from SSO (AAD login page in your case) **if the user doesn't exist in your db.**

`var emailAddress = info.Principal.FindFirstValue(AbpClaimTypes.Email);`
Could you be able to debug this line down to user creation?

**Comment by Znow** [comment]
@gterdem 

From my exhibit in previous comment, the method is not being debugged at all, no breakpoint are being hit.

These are my steps:

1. Start project and run website
2. Access /account/login
3. Microsoft login page shows
4. Enter credentials, and the click "Login", below is being shown
![image](https://user-images.githubusercontent.com/265719/75116363-d85fd200-5667-11ea-8b9f-6352c8595d25.png)

**Comment by Znow** [comment]
@gterdem The fix that you did solved the issue with not being able to login.

But now a new problem have arisen. 

1. Run project
2. Azure AD (Microsoft) login page shows
3. Login with AD credentials
4. Logged in

So far so good.

But, I should actually be presented with the "/Account/Login"-page, shouldn't I?

Because when I just get prompted for login with the Microsoft login page, now Identity user is created in the database, at all...

**Comment by Znow** [comment]
@gterdem @ebicoglu @hikalkan bump - any update on this?

**Comment by yekalkan** [comment]
@Znow 

- have you marked the new Login.cshtml as embedded resource?

```
  <ItemGroup>
    <Content Remove="Pages\Account\Login.cshtml" />
  </ItemGroup>

  <ItemGroup>
    <EmbeddedResource Include="Pages\Account\Login.cshtml" />
  </ItemGroup>
```

- have you set virtual file system configration?

```
            Configure<AbpVirtualFileSystemOptions>(options =>
            {
                options.FileSets.AddEmbedded<MyProjectNameWebModule>("MyCompanyName.MyProjectName.Web");
            });
```

**Comment by Znow** [comment]
@yekalkan - Same stuff, not hitting breakpoints in Login.cshtml.cs

**Comment by yekalkan** [comment]
![image](https://user-images.githubusercontent.com/5793291/76628403-b7dfb500-654d-11ea-9a6e-2ba7ed34214b.png)

I'm able to debug, following the steps above.

**Comment by Znow** [comment]
@yekalkan Okay, and what are your steps?

I am not hitting any breakpoints?

![image](https://user-images.githubusercontent.com/265719/76638768-4b64b600-654d-11ea-8058-7efdf4fdd180.png)

![image](https://user-images.githubusercontent.com/265719/76638793-561f4b00-654d-11ea-85b9-04499b7a734e.png)

**Comment by yekalkan** [comment]
![image](https://user-images.githubusercontent.com/5793291/76646191-9d1c3900-656b-11ea-888f-2a4f21cd3cb7.png)


![image](https://user-images.githubusercontent.com/5793291/76646171-91c90d80-656b-11ea-9658-f72e5e057db5.png)

Same as yours. Does your project have MyCompanyName as prefix? If so, it is missing in `AddEmbedded<MyProjectNameWebModule>("MyCompanyName.MyProjectName.Web")`. Check it, just in case.

**Comment by Znow** [comment]
Nope it hasn't - it's just "X.Web" etc.

**Comment by Znow** [comment]
@yekalkan - Reading up on https://docs.abp.io/en/abp/latest/UI/AspNetCore/Customization-User-Interface told me that I should have my Login.cshtml.cs as below:

```
[Dependency(ReplaceServices = true)]
    [ExposeServices(typeof(LoginModel))]
    public class MyLoginModel : LoginModel
    {
```

But with this, I get the error below when accessing "/Account/Login" (still no breakpoints being hit):
![image](https://user-images.githubusercontent.com/265719/76678060-fd939080-65d4-11ea-8938-f72b641e2c66.png)

**Comment by Znow** [comment]
@yekalkan @gterdem @hikalkan Another observation - I cannot login with the standard local "admin" user when AAD login is enabled.... This seems to be a bug aswell, even if I have applied:

Appsettings.json
```
  "Settings": {
    "Abp.Account.EnableLocalLogin": "true" 

  }
```

**Comment by gterdem** [comment]
@Znow Hello, sorry couldn't look into anything for sometime because of RL issues.

The problem is, there is no `ClaimTypes.NameIdentifier` (Microsoft Claim type of `http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier`) is provided by AzureAD. That's why you are getting `SignInManager.GetExternalLoginInfoAsync()` null.

Try overriding SignInManager and change the `ClaimTypes.NameIdentifier` to `sub`

https://github.com/aspnet/Identity/blob/feedcb5c53444f716ef5121d3add56e11c7b71e5/src/Identity/SignInManager.cs#L611

It should fix your problem.

**Comment by Znow** [comment]
@gterdem Hi, thanks for still looking into this.

Yeah, I got so far that I'm overriding SignInManager, and can debug GetExternalLoginInfoAsync(), but "auth" doesnt get me further:

![image](https://user-images.githubusercontent.com/265719/76882113-78c9a080-687a-11ea-98e0-6916e2333039.png)

**Comment by gterdem** [comment]
@Znow Nice to hear you got it moving.
It is signing in but authenticating against wrong scheme after user created. You may want to look into that.

I will try to reproduce the whole flow once I get time for it. It would be nice if you can put some code on a public repository when you are done so that we can troubleshoot better and help better.

**Comment by Znow** [comment]
@gterdem Ah, that enlightens some more, thanks! I can ofcourse create a public repo during today, thanks.

**Comment by Znow** [comment]
@hikalkan Why is it removed from 2.3?

**Comment by hikalkan** [comment]
* 2.3 will be released tomorrow.
* I don't see any development should be done for this issue yet. The discussion and investigation in progress. So, why we wait 2.3 release?

**Comment by Znow** [comment]
@gterdem I get signed in now with my Azure AD account - but ABP doesn't see me as signed in. I can see that there are cookies etc.

![image](https://user-images.githubusercontent.com/265719/77072217-a420cc80-69ed-11ea-90fe-99f56002b4d6.png)

**Comment by Znow** [comment]
@gterdem Have you had a chance to look at the repo I created?

**Comment by gterdem** [comment]
Hello @Znow,
I sent a PR to your [repository](https://github.com/Znow/AbpAzureAdLogin).
There are some problems with setting `CurrentUser` authenticated status while `HttpContextUser` is authenticated just fine. 

I need to investigate further about schemes to find a viable solution to this AD External Login. 
Thanks for your patience.

**Comment by Znow** [comment]
@gterdem Yeah I saw it, and it's merged in already =)

Thanks alot for your help so far, really appreciate it.

**Comment by Znow** [comment]
@hikalkan - @gterdem was so kind and "solved" the issue, in the repository on my account, which I linked to previously, its a working solution now.

So I think this issue can be closed, if there are no stuff to be fixed in ABP.

Thanks

**Comment by gterdem** [comment]
@Znow Thanks Znow, I am glad if i could help. We fixed a problem on our side about guid generation. There will be blogpost about detailed AzureAD integration for abp Mvc app soon to help others having trouble with integration.
I'll edit this post when blogpost is released.

**Comment by Znow** [comment]
@gterdem I have deployed the app to Azure, and cannot get it to work. It throws a HTTP 500.

I deployed the test repository I created, which you forked and created PR for.

![image](https://user-images.githubusercontent.com/265719/81283415-e022f600-905c-11ea-94ea-8da33b536bee.png)

```
2020-05-07 10:15:21.460 +00:00 [ERR] An unhandled exception has occurred while executing the request.
System.ArgumentNullException: Value cannot be null. (Parameter 'userName')
   at Volo.Abp.Check.NotNull[T](T value, String parameterName)
   at Volo.Abp.Identity.IdentityUser..ctor(Guid id, String userName, String email, Nullable`1 tenantId)
   at Volo.Abp.Account.Web.Pages.Account.LoginModel.CreateExternalUserAsync(ExternalLoginInfo info)
   at Volo.Abp.Account.Web.Pages.Account.LoginModel.OnGetExternalLoginCallbackAsync(String returnUrl, String returnUrlHash, String remoteError)
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.ExecutorFactory.GenericTaskHandlerMethod.Convert[T](Object taskAsObject)
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.ExecutorFactory.GenericTaskHandlerMethod.Execute(Object receiver, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.PageActionInvoker.InvokeHandlerMethodAsync()
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.PageActionInvoker.InvokeNextPageFilterAsync()
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.PageActionInvoker.Rethrow(PageHandlerExecutedContext context)
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.PageActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.PageActionInvoker.InvokeInnerFilterAsync()
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextExceptionFilterAsync>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ExceptionContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Volo.Abp.AspNetCore.Serilog.AbpSerilogMiddleware.InvokeAsync(HttpContext context, RequestDelegate next)
   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.<>c__DisplayClass5_1.<<UseMiddlewareInterface>b__1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Volo.Abp.AspNetCore.Auditing.AbpAuditingMiddleware.InvokeAsync(HttpContext context, RequestDelegate next)
   at Volo.Abp.AspNetCore.Auditing.AbpAuditingMiddleware.InvokeAsync(HttpContext context, RequestDelegate next)
   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.<>c__DisplayClass5_1.<<UseMiddlewareInterface>b__1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)
   at Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)
   at Microsoft.AspNetCore.Localization.RequestLocalizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.RequestLocalization.AbpRequestLocalizationMiddleware.InvokeAsync(HttpContext context, RequestDelegate next)
   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.<>c__DisplayClass5_1.<<UseMiddlewareInterface>b__1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at IdentityServer4.Hosting.IdentityServerMiddleware.Invoke(HttpContext context, IEndpointRouter router, IUserSession session, IEventService events)
   at IdentityServer4.Hosting.MutualTlsTokenEndpointMiddleware.Invoke(HttpContext context, IAuthenticationSchemeProvider schemes)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at IdentityServer4.Hosting.BaseUrlMiddleware.Invoke(HttpContext context)
   at Volo.Abp.AspNetCore.MultiTenancy.MultiTenancyMiddleware.InvokeAsync(HttpContext context, RequestDelegate next)
   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.<>c__DisplayClass5_1.<<UseMiddlewareInterface>b__1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Builder.ApplicationBuilderAbpJwtTokenMiddlewareExtension.<>c__DisplayClass0_0.<<UseJwtTokenMiddleware>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>g__Awaited|6_0(ExceptionHandlerMiddleware middleware, HttpContext context, Task task)
```

**Comment by gterdem** [comment]
@Znow hello,

Can you create a new issue explaining with reproduce steps.

Thanks.
